### PR TITLE
Rework text replacements to allow %#.variable

### PR DESCRIPTION
**Pull request by InfusOnWoW**
Where # is a trigger number.

### CODE REVIEW

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 1]
```diff
@@ -2646,18 +2716,104 @@ local function flattenRegionOptions(allOptions, data, isGroupTab)
...
+    end
+  end
+end
+
+local function AddSubRegion(data, subRegionName)
+  WeakAuras.ApplyToDataOrChildData(data, AddSubRegionImpl, subRegionName)
+  WeakAuras.RefillOptions()
+end
+
+local function AddOptionsForSupportedSubRegion(regionOption, data, supported)
```
Comment (emptyrivers) [inline WeakAurasOptions/WeakAurasOptions.lua]: If there is no supported subregion, then this should not be called, I think.

**Comment by Stanzilla** [comment]
What's the status here? I guess we'll move it to 2.12 anyway, right?

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 2]
```diff
@@ -2525,18 +2494,105 @@ local function flattenRegionOptions(allOptions, isGroupTab)
...
+  WeakAuras.ApplyToDataOrChildData(data, AddSubRegionImpl, subRegionName)
+  WeakAuras.ReloadOptions2(data.id, data)
+end
+
+local function AddOptionsForSupportedSubRegion(regionOption, data, supported)
+  local result = {}
+  local order = 1
+  result.__order = 300
+  result.__title = L["Add Sub Regions"]
+  for subRegionType in pairs(supported) do
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: This adds a button per potential sub region type. That'll look horrible with more than 2 potential sub region types, but we can figure that out once there's more than 1 :)

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 3]
```diff
@@ -2658,6 +2717,46 @@ function WeakAuras.EnsureOptions(id)
...
+scheduleRefillFrame.ScheduleReload = function(self, id, data)
+  scheduleRefillFrame.ids[id] = data
+  scheduleRefillFrame:ScheduleRefillOnly()
+end
+
+scheduleRefillFrame.ids = {}
+
+
+-- TODO replace older ReloadOptions, SchduleReloadOptions, ReloadTriggerOptions, ReloadGroupRegionOptions
+-- automatically clear parent/tempGroup ?
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: This is something that I want to tackle soon, and will rework how we treat the options.

Basically, we sometimes need to regenerate the options table, and sometimes just tell AceGui that our options changed even without the table being changed.

Currently we do that by immediately recreating a new options table. What I want to try out, is to split that up into two steps:

a) Just mark which options tables need to be recreated / mark that options need to be refilled
b) In a OnUpdate handler recreate the options table.

This should ensure that each change only recreates each table once at most. I have to test and remove lots of old code, but that'S the basic idea. I think that should make editing big groups much faster.

**Comment by InfusOnWoW** [comment]
A few general comments on the PR.

** Trigger to Region Communication
This changes how the WeakAuras triggers communicate with the regions. Previously this was done via a set of methods, e.g. SetDurationInfo, SetName, SetStacks etc, which were called with the active state's information. Sub texts support text replacement from arbitary triggers, thus they need to be informed if the any triggers changed, not just the currently active one. This is similar to conditions, that also work on all triggers.

Thus, this patch introduced a new "Update()" method, that is on each region/sub region and is called whenever the any state or which state is active changed.

SetDurationInfo does continue to work, but is no longer actually used internally.

The longer term idea is, that this also empowers regions to e.g. show progress from a secondary trigger.

** Region vs sub region design
The sub regions / regions design:
* A region does not know of its sub regions.
* But a sub region knows the type and parameters of its parent.

That second part helps with some of the idea for further sub regions, e.g. a way to position a "divider" at a position on progress bars. That requires that sub region to know what the exact configuration of the progress bar is. 

** Anchoring
* Each region declares the potential anchors via a GetAnchors and provides a AnchorSubRegion function that can anchor a sub region to a anchor. A region can provide two types of anchors: "point" anchors and "area" anchors. E.g. a aurabar provides the 9 points of the actual bar, the 9 points of the icon and the spark position. And for area, should provide: bar, icon and all.

[The area options are untested, there's some code for them, but not 100% finished. This needs to be done once a sub region makes use of that]

A sub region controls the anchor options it shows, e.g. the above mentioned dividing line would not offer any option. For sub texts a self point option is offered, which defaults to a "automatic" mode.

** Frame Levels
Currently each sub region simply gets a automatically increasing frame level. This is good enough for texts, but other sub region types will want to customize that.

** Options code
Instead of the old code that gave fake information via SetDurationInfo, on opening the options fake states are created. These are created by the trigger systems, and the GenericTrigger forwards that to custom triggers with a fake "OPTIONS" event. So custom triggers can now control what information is shown while the options are open. 
This fake event, might break a few custom triggers though.

File: WeakAurasOptions/WeakAurasOptions.toc [hunk 4]
```diff
@@ -29,6 +29,8 @@ RegionOptions\DynamicGroup.lua
 RegionOptions\Model.lua
 RegionOptions\ProgressTexture.lua
 
+SubRegionOptions\SubText.lua
```
Comment (emptyrivers) [inline WeakAurasOptions/WeakAurasOptions.toc]: Add this file to babelfish, too.

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 5]
```diff
@@ -2464,18 +2382,69 @@ local function addCollapsibleHeader(options, key, title, order, isGroupTab)
...
-    return WeakAuras.IsCollapsed("collapse", "region", key, false)
+
+  if hasUp then
+    options[key .. "upButton"] = {
+      type = "execute",
+      name = "",
+      order = order + 0.3,
+      width = 0.15,
+      func = input.__up,
+      image = "Interface\\AddOns\\WeakAuras\\Media\\Textures\\moveup",
```
Comment (emptyrivers) [inline WeakAurasOptions/WeakAurasOptions.lua]: It occurs to me that we could might do well to make this less magic in the same way that we did for new feature indicators

File: WeakAurasOptions/WeakAurasOptions.lua (hunk 3 above)
Comment (emptyrivers) [reply WeakAurasOptions/WeakAurasOptions.lua]: Can you speak about that in #821?

**Comment by emptyrivers** [comment]
Oh, also it would be good to document the new "OPTIONS" subevent. People seem to clamor about something for that (for admittedly insane reasons), so since it exists we may as well tell people about it.

File: WeakAuras/WeakAuras.lua [hunk 6]
```diff
@@ -3172,6 +3507,20 @@ function WeakAuras.PreAdd(data)
   if default then
     WeakAuras.validate(data, default)
   end
+
+  if data.subRegions then
+    for index, subRegionData in ipairs(data.subRegions) do
+      local subType = subRegionData.type
+      local default = type and WeakAuras.subRegionTypes[subType] and WeakAuras.subRegionTypes[subType].default
```
[the remark this answers is not in the document]
Comment (InfusOnWoW) [reply WeakAuras/WeakAuras.lua]: * Fixed the typo
* Well, I don't want to name it "text", as I think that'll lead to confussion with the other text, so I disagree that this is something that should change.
* All sub regions need to be in the same list, as the ordering between them affects frame levels. And I actually imagine that mattering. Now, somewhat confussingly, some sub regions won't care about frame levels and thus their position in the list is not important.

File: WeakAuras/WeakAuras.lua [hunk 7]
```diff
@@ -4699,14 +5035,34 @@ local function ApplyStatesToRegions(id, triggernum, states)
   for cloneId, state in pairs(states) do
     if (state.show) then
       local region = WeakAuras.GetRegion(id, cloneId);
-      if (not region.toShow or state.changed or region.state ~= state) then
-        ApplyStateToRegion(id, cloneId, region, state, parent);
+      local applyChanges = not region.toShow or state.changed or region.state ~= state
+      region.state = state
+      region.states = region.states or {}
+      local needsTimerTick = false
+      for triggernum = -1, triggerState[id].numTriggers do
```
[the remark this answers is not in the document]
Comment (InfusOnWoW) [reply WeakAuras/WeakAuras.lua]: -1 is the global state for global conditions
0 is always nil

File: WeakAurasOptions/WeakAurasOptions.lua (hunk 5 above)
Comment (InfusOnWoW) [reply WeakAurasOptions/WeakAurasOptions.lua]: I'm not yet bothered enough to do that :)

**Comment by emptyrivers** [comment]
I have found a bug:

1. Create an aura which supports text subregions (in this example, icon).
2. Go to conditions, check the property dropdown. It should look like this: 
image
3. Go to display, click "Add Text". Go back and look at the property dropdown in Conditions, it should now look like this (note the third text option)
image
4. Go back to display, and delete one text subregion. Go back to conditions, and look yet again at the conditions dropdown. Note that the 3. Text property still exists in the dropdown:
image
5. If you attempt to use this property, then you will (somewhat predictably) get an error which looks similar to this:
```
2x [string "--[[ Error in 'New' ]] local newActiveCondi..."]:38: attempt to index field '?' (a nil value)
[string "--[[ Error in 'New' ]] local newActiveCondi..."]:38: in function `?'
WeakAuras\WeakAuras-@project-version@.lua:1167: in function <WeakAuras\WeakAuras.lua:1161>
WeakAuras\WeakAuras-@project-version@.lua:1180: in function <WeakAuras\WeakAuras.lua:1173>
[C]: in function `CameraOrSelectOrMoveStop'
[string "CAMERAORSELECTORMOVE"]:4: in function <[string "CAMERAORSELECTORMOVE"]:1>

Locals:
region = <unnamed> {
 SoundRepeatStop = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:156
 glow = false
 SetUseGlowColor = <function> defined @WeakAuras\RegionTypes\Icon.lua:480
 AnchorSubRegion = <function> defined @WeakAuras\RegionTypes\Icon.lua:149
 xOffset = 0
 Expand = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:580
 yOffset = 0
 anchorPoint = "TOP"
 GetColor = <function> defined @WeakAuras\RegionTypes\Icon.lua:391
 inner = <unnamed> {
 }
 icon = <unnamed> {
 }
 SetOffset = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:264
 keepAspectRatio = false
 SendChat = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:219
 SetInverse = <function> defined @WeakAuras\RegionTypes\Icon.lua:432
 toShow = true
 SetRegionHeight = <function> defined @WeakAuras\RegionTypes\Icon.lua:427
 useAuto = true
 OptionsClosed = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:521
 glowType = "buttonOverlay"
 Color = <function> defined @WeakAuras\RegionTypes\Icon.lua:363
 cloneId = ""
 Update = <function> defined @WeakAuras\RegionTypes\Icon.lua:584
 regionType = "icon"
 stacks = <unnamed> {
 }
 SetIcon = <function> defined @WeakAuras\RegionTypes\Icon.lua:398
 relativePoint = "TOP"
 SoundPlay = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:204
 zoom = 0
 color_b = 1
 UpdateTexCoords = <function> defined @WeakAuras\RegionTypes\Icon.lua:298
 alpha = 1
 color_r = 1
 SetXOffset = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:273
 SetGlow = <function> defined @WeakAuras\RegionTypes\Icon.lua:494
 SetRegionWidth = <function> defined @WeakAuras\RegionTypes\Icon.lua:422
 SetGlowType = <function> defined @WeakAuras\RegionTypes\Icon.lua:459
 RunCode = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:226
 GetYOffset = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:285
 values = <table> {
 }
 cooldownSwipe = true
 GetXOffset = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:281
 cooldownEdge = false
 SetCooldownEdge = <function> defined @WeakAuras\RegionTypes\Icon.lua:446
 color_g = 1
 text2 = <unnamed> {
 }
 SetAnchor = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:251
 ColorAnim = <function> defined @WeakAuras\RegionTypes\Icon.lua:377
 SetRegionAlpha = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:298
 states = <table> {
 }
 state = <table> {
 }
 subRegions = <table> {
 }
 color_a = 1
 SetZoom = <function> defined @WeakAuras\RegionTypes\Icon.lua:454
 Collapse = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:563
 SetOffsetAnim = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:289
 ClickToPick = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:526
 PreShow = <function> defined @WeakAuras\RegionTypes\Icon.lua:520
 frameId = "new"
 scaley = 1
 glowStop = <function> defined @WeakAuras\Libs\LibCustomGlow-1.0\LibCustomGlow-1.0.lua:675
 xOffsetAnim = 0
 yOffsetAnim = 0
 glowStart = <function> defined @WeakAuras\Libs\LibCustomGlow-1.0\LibCustomGlow-1.0.lua:592
 glowColor = <table> {
 }
 SetYOffset = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:277
 PreShowGlow = <function> defined @WeakAuras\RegionTypes\Icon.lua:520
 SetGlowColor = <function> defined @WeakAuras\RegionTypes\Icon.lua:487
 SoundStop = <function> defined @WeakAuras\RegionTypes\RegionPrototype.lua:165
 0 = <userdata>
 SetAnimAlpha = <function> defined
```

**Comment by mrbuds** [comment]
Text rotation doesn't show correctly: https://i.imgur.com/sJpWojA.gif

```
!nsvZUjomq4NM9WEOsqcfwosPHs1Ygq40vTxmYK4KyvJD2yhOTp97m2jqA7HwjKzY4pp)(ndDinHse04HHxpD0akXqj3YpsjhOKm8hCBHuFIUNsEzDEUHBPXaSx7jZuPL66nAHYsjZJItI2sjPATmtFsrojQ40BUOikRGJwJLAfALbDGXYQTGa8CHs0kLdIMsNm8LTwuuWRr4XUqUvbk2y4XSdoNWAQzkqUfhz4e8X8JCm0wYzsBjL0bauzQ4s59zDF1S3bL0KNlEHs2nFgjzhjz2w0lyKrQKSxbVEg6MAUdkzt0Qvqu9kKSUOiakE89GHsCQwgTAZIhwHEPrDj2D5gwjoYt8k)JoJtJVA4aF1WYRvm5FHmhkwqknMsEMZRMbbEQDldkHUAPsCaLv9RMTbJsRGZmiKqa78kn8uTcsBWhhyc13h95UY3dVfRcgMfV311rU0CTux7Bq4Va3zO7CeE67eB5fT8J2go)f7otjd4qp2r(cBvJhuYpQEhQp5MfJUoyyW4b)Aqq4KPtMgqZV66qNRhCX1ElWL5Te6zpKSUJJ)upwFT2c5uIZZBV)ULNljqS7JhNL8p8XEp0Po9RRboC5AmgUrYEJN9ol(0hT4rHrSx6Mc6bRnjw)amuUlz9MEM1tmJDnWZkjI3a6xW4Um(tbUV2E27b(MCjxuucFoEeLi1m3AdKII)FOrAfDC9ujZy(Kwd61pO06jl(OCFJ1QvRpYRH5pO274g(ReaxR90TL5lRRi3u0T7X2LOlQHTcElsMVnkkMsEtRp4xW1y1(Tlwa56QEpwaDLy(jyQgLE8Ur)EHv88YF(paTSQK52cDsKb7DCfNUTGiV5wHHbDSmCUacQCrr3oqipnUXLCmQGDDmR3sWQUltqW9zcdUr6(u3YHWWPtgEXhiGEe52nZG0ra7OaSyq))p
```

**Comment by mrbuds** [comment]
Clicking the "X" to remove a sub region raise an error
```
3x [string "--[[ Error in 'New' ]] local newActiveCondi..."]:38: attempt to index field '?' (a nil value)
[string "--[[ Error in 'New' ]] local newActiveCondi..."]:38: in function `?'
WeakAuras\WeakAuras-@project-version@.lua:5064: in function <WeakAuras\WeakAuras.lua:5025>
WeakAuras\WeakAuras-@project-version@.lua:5146: in function `UpdatedTriggerState'
WeakAuras\WeakAuras-@project-version@.lua:4915: in function `UpdateFakeStatesFor'
WeakAuras\WeakAuras-@project-version@.lua:3611: in function <WeakAuras\WeakAuras.lua:3533>
WeakAuras\WeakAuras-@project-version@.lua:3625: in function `Add'
WeakAurasOptions\SubRegionOptions\SubText.lua:12: in function `func'
WeakAuras\WeakAuras-@project-version@.lua:118: in function `ApplyToDataOrChildData'
WeakAurasOptions\SubRegionOptions\SubText.lua:75: in function <WeakAurasOptions\SubRegionOptions\SubText.lua:74>
[C]: ?
...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0-69.lua:50: in function <...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0.lua:48>
...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0-69.lua:789: in function <...nfig-3.0\AceConfigDialog-3.0\AceConfigDialog-3.0.lua:610>
[C]: ?
AzeriteTooltip\libs\AceGUI-3.0\AceGUI-3.0-36.lua:71: in function <AzeriteTooltip\libs\AceGUI-3.0\AceGUI-3.0.lua:69>
AzeriteTooltip\libs\AceGUI-3.0\AceGUI-3.0-36.lua:285: in function `Fire'
...ooltip\libs\AceGUI-3.0-36\widgets\AceGUIWidget-Icon.lua:26: in function <...ooltip\libs\AceGUI-3.0\widgets\AceGUIWidget-Icon.lua:25>
```

**Comment by mrbuds** [comment]
from what was said on discord i though the following aura with 2 auras trigger w/ auto-clone would work
```
!Dx1wVPTpu8pntA7HnrsPDBpsPb1krtO40PUnndMeNG3m2iBhOTpWN9)NJD4pxQuV8WEzsOKdN87C)gnIMtjcAA0jN(1UDOelLCbFfLSGskXFWxRL610zuY9zvvwUJMcWEypAMQyU2msluokPFsAEYykPqRLL61kYAXso98DmskR5O2yfoHwzrdyDmJdiaXRekHDEGgidCbANruxZni8uVl3YWl(C96(sTIBr70y5PSfEtYAmmfq3kfj6ZOQ4R4OJEjNjDZbPBM55mYWRe3tjKrjdhcSxYLYRkTbxzlistLh0K(9i5ti59gd2Vb9tsTr3SKs2Ap0NFaICVtedzs(mq2CpRltgoAWTdrmnQDbcgOPXpBWDGkpo8IJo9S43qe(SX0US4bH3RmmiLc7VBuqjEfaSOX60l2(opGCOUwuafCpkT69UpSzQH7AmQntD)m6xBMYvLH2Kv8wzUwxYPPFmQtO9WXnkM8BqBbOaifCgL8hoFzpO2v4gZa1InAwUS6OMZsULbwI58DIy)DFTuBc5r8xS)5j(NDXNHEGX862E22Mq(9Uj25mOV(U9gicvjap(9wubY3f9PLTmkEzl6XvPrpFGr84MP30WkbVMTzA(HwFF3VJxzD8kRZrklG(777RE27sr9UnpBplhQ0z3Mp8Q0Kw(hmVdFkz8K8SrB3d8eLVsyfZK(wjJ2b58CFQindvyqM7owg00eXJqXokE3uXBoBh)Vz2(8S88SR)BMWb4Z5I65aIZ6sjsnZFjahSW3lAKoX2r9cjZAFcxlQSJy6cZAHOCwJZPvzR4gj7bWt9JwHpjkGXz4GHQsuhwLYuIf48SA)ZfKsCegyoPTfGdIaBSB7iuWstu2fmH61J()pa96W72DydZYxiSmi1x6pW140(Rq(m(adSoneEK(JtssHWewVLYx7x2UBheizfIfoca9E(f7RfLWECFLWNAWLYOW3uK0B8GV8JrzNGRdHuPL3AzWIzl3EGfCYh1WcySAJbHyVpWKlNhSYlpIaWxbuDJX)q)Vd
```
I expected it to show progress remaining for both "power word: shield", and "power word: fortitude" on myself, and it show only the power word shield

**Comment by InfusOnWoW** [comment]
@Buds I can't reproduce the error message on removing a sub text.

On your group trigger you need to enable the "combine matches per unit", since only with that option enabled there can be just one aura per group member and only in that case the guid is used.

I fixed the problem rivers found and the text rotate.

**Comment by mrbuds** [comment]
I can't reproduce the error on removing sub text either now

**Comment by InfusOnWoW** [comment]
On rebasing I had the obvious merge conflicts, but wrongly resolved them. I've squashed the commits except the new commit because I couldn't be bothered to redo the rebasing and merge resolving.

**Comment by emptyrivers** [comment]
@InfusOnWoW you have some merge conflicts to fix btw

**Comment by InfusOnWoW** [comment]
I'm not too concerned about merge conflicts at the moment, they shouldn't be too hard to resolve once we have the next release out.

**Comment by InfusOnWoW** [comment]
Something in that rebase went wrong, I'll need to redo it.

**Comment by InfusOnWoW** [comment]
Rebased again, should be ready now.

**Comment by Causese** [comment]
aura: https://wago.io/oxD23fi-3

**Comment by Causese** [comment]
- auras that used to be "outer - bottom" are "inner - bottom right" after installing the branch
https://wago.io/bfaraid2 for example

- non aura-triggers don't show the "expiration time" preview anymore.
live: https://i.gyazo.com/fbd9ee4fa0054207cac5b01ecbf79065.png
branch: https://i.gyazo.com/1f1acaf2486f46add63159e60a4aac77.png

- https://wago.io/bfaraid2 causes an error upon import that only stops when unticking dynamic duration in trigger.
1x [string "--[[ Error in '05 Krag'wa's Wrath' ]] retur..."]:2: attempt to perform arithmetic on field 'expirationTime' (a nil value)

**Comment by Causese** [comment]
!(https://i.gyazo.com/31f50dd9c56b42fdb65f5e52b39971b1.png)
aura: https://wago.io/Mp6dK6_GG

**Comment by Causese** [comment]
backup SV before:

1. create icon,
2. convert to text
`1x WeakAuras\SubRegionTypes\SubText.lua:319: attempt to call method 'AnchorSubRegion' (a nil value)`

cant /wa after following those steps and reloading
4x WeakAuras\WeakAuras-@project-version@.lua:74: attempt to call field 'ToggleOptions' (a nil value)

**Comment by InfusOnWoW** [comment]
Is %c with trigger update firing every frame intended.

**Comment by InfusOnWoW** [comment]
@Causese Thanks, I've fixed all the issues you found except that I couldn't reproduce:
"auras that used to be "outside - bottom" are "inner - bottom right" after installing the branch"

As far as I can see, the outside bottom text are still on the outside and bottom with my branch.

**Comment by Causese** [comment]
you're right. it's still "bottom outer" but "buttom - outer" has to be re-applied for it to actually be "bottom - outer".

e.g change it to "outer - bottom right" and then back to bottom

**Comment by InfusOnWoW** [comment]
@Causese can you add a screenshot how it looks like after importing?

**Comment by Causese** [comment]
live: https://i.gyazo.com/fbd9ee4fa0054207cac5b01ecbf79065.png
branch: https://i.gyazo.com/1f1acaf2486f46add63159e60a4aac77.png

**Comment by Causese** [comment]
Auras that had their 2. Text unticked, now have a 2. Text with text rewrite branch.

you can try wago.io/bfaraid2 where many auras were added 2. text after installing the branch & importing it

**Comment by InfusOnWoW** [comment]
I can't reproduce that, after importing https://wago.io/bfaraid2 the texts are at the bottom and not in the middle as in your screenshots.

Also I tested a few auras that had the second text is disabeld e.g.: "Biding Faith", "Call to Arms", "Divine Mallet" and just like expected the second text for those is not visible.

**Comment by Causese** [comment]
**backup before**
upon importing https://wago.io/ZqrTSehX6:

```
1x WeakAuras\WeakAuras-@project-version@.lua:5433: attempt to index local 'regionState' (a nil value)
WeakAuras\WeakAuras.lua:5433: in function <WeakAuras\WeakAuras.lua:5417>
WeakAuras\WeakAuras.lua:5513: in function `ReplacePlaceHolders'
...rfaceWeakAurasOptions\RegionOptions\Text.lua:207: in function `UpdateText'
...rfaceWeakAurasOptions\RegionOptions\Text.lua:220: in function `modifyThumbnail'
WeakAurasOptions\WeakAurasOptions-@project-version@.lua:4626: in function `SetThumbnail'
WeakAurasOptions\WeakAurasOptions.lua:4591: in function `UpdateDisplayButton'
WeakAurasOptions\WeakAurasOptions.lua:1581: in function <WeakAurasOptions\WeakAurasOptions.lua:1575>
```

**Comment by Causese** [comment]
import: https://wago.io/ZqrTSehX6
1) left-click "Cannon Sequence (Air Horn)" and notice how it vanishes after it's timed value
2) "RING OF ICE BAR" has a smaller font on text rewrite branch (ptr) compared to 2.12.4.1(live)
3) %i returns questionmark's file path instead of an icon in a text aura

**Comment by Causese** [comment]
import: https://wago.io/bfaraid1timers
2. text is visible and uses %p when importing it to branch, it's not enabled when importing it to 2.12.4.1
expected it to atleast not be visible

**Comment by Causese** [comment]
import: https://wago.io/buffchecker
2.11.4.1: https://i.gyazo.com/61afc59e581d076c39803347ba730de9.png
branch: https://i.gyazo.com/c89ecead43855611f36086478cc00d72.png
%i and other replacements show %i/%p etc next to the text instead of the actual icon etc.

also odd behaviour on branch: 
https://i.gyazo.com/f3226aa9d0d414d3953c00941efecdf9.gif
clicking the eye doesn't make all texts show

**Comment by Causese** [comment]
%s shows 1 for both bars and icons on master and nothing on text rewrite (aura - player)

**Comment by Causese** [comment]
import: https://wago.io/H1XSZ9TH7
`1x WeakAuras\GenericTrigger.lua:673: attempt to compare number with nil
`

**Comment by Causese** [comment]
import: https://wago.io/bfaraid3
**on 2.12.4.1:** 02 Gift of N'Zoth Energy doesn't have a questionmark icon upon import
**branch:** Gift of N'Zoth Energy has a questionmark icon upon import

**Comment by Causese** [comment]
aura's that had a condition and text 1 - color - alpha as property change (2.12.4.1) are blank with branch
try https://wago.io/M28UoTFSh for example

**Comment by Causese** [comment]
Create a new bar and immediately convert it to icon
`3x WeakAuras\RegionTypes\Icon.lua:247: <unnamed>:SetPoint(): Unknown region point`

**Comment by InfusOnWoW** [comment]
@Causese Gift of N'Zoth having no icon is expected now. There's no a fake OPTIONS event sent to a aura, to which the aura should respond with what it wants to be shown in the options.

**Comment by InfusOnWoW** [comment]
On https://wago.io/M28UoTFSh I get a " Error: page_not_accessible "

**Comment by Causese** [comment]
odd behaviour: https://i.gyazo.com/bfcef5e836f244f0ad3d25549c913a8d.gif
import: https://wago.io/buffchecker

**Comment by Causese** [comment]
`1x ...aceWeakAuras\RegionTypes\RegionPrototype.lua:214: attempt to index local 'options' (a nil value)`

**Comment by Causese** [comment]
display doesn't show "anchor-related-display-settings" when selecting group and visiting display tab
to reproduce just make a dynamic group with 2 icons, select the group and visit display tab.

**Comment by Causese** [comment]
text is anchored left/center-ish instead of bottom right.
duplicating/importing the aura fixes it therefore u need my SV to reproduce.

[WeakAuras.zip](https://github.com/WeakAuras/WeakAuras2/files/3277646/WeakAuras.zip)

**Comment by Causese** [comment]
condition fire when clicking the aura while in options so it could probably also use the WeakAuras.IsOptionsOpen() treatment that you recently added to trigger.

To reproduce:
1. make an icon aura
2. add condition if active true then run custom code
3. add SendChatMessage("Hi" , "SAY") in custom code field
4. every time u select the aura and it's visible, the message will fire.

this does not happen on 2.12.4.2

**Comment by Causese** [comment]
the change you did to stop trigger update from spamming also has to be done for text region.
to reproduce:
1. Make a text type aura
2. add %c with following code
```
function()
    if aura_env.state then
print("spam")
    end
end

```

**Comment by Causese** [comment]
Deleting big groups with text-rewrite branch causes
`1x ...aceWeakAuras\RegionTypes\RegionPrototype.lua:87: script ran too long`
and only deletes about 20 auras per attempt to delete the entire group, and causes the error for each attempt until 117 ish auras are left.

it works in one go on 2.12.4.2

example aura:
https://wago.io/kR7bGhWG5

**Comment by InfusOnWoW** [comment]
I can't reproduce the problem with the Ice Barrier aura. Even with your settings the text is in the bottom right.

Yes, we now run the custom text functions while the options are open. That's a intended change. If you use print in there, you get what you asked for...

I've made deleting groups faster so that the script run to long should not appear.

**Comment by InfusOnWoW** [comment]
Also rebaed now.

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 8]
```diff
@@ -1293,15 +1293,15 @@ function WeakAuras.DeleteOption(data)
   thumbnails[id] = nil;
   displayButtons[id] = nil;
 
-  if(parentData and parentData.controlledChildren) then
+  if(parentData and parentData.controlledChildren and not massDelete) then
```
Comment (emptyrivers) [inline WeakAurasOptions/WeakAurasOptions.lua]: cute

**Comment by InfusOnWoW** [comment]
Text placement and custom functions in trigger update mode for text auras is fixed.

File: WeakAuras/RegionTypes/Icon.lua [hunk 9]
```diff
@@ -265,6 +265,31 @@ local function create(parent, data)
...
+
+    if MSQ then
+      iconWidth = region.button:GetWidth()
+      iconHeight = region.button:GetHeight()
+    else
+      iconWidth = region:GetWidth()
+      iconHeight = region:GetHeight()
+    end
+
+    print("UpdateInnerOuterSize ", iconWidth, iconHeight)
```
Comment (Stanzilla) [inline WeakAuras/RegionTypes/Icon.lua]: todo

**Comment by Causese** [comment]
%p from text auras doesn't update anymore, stuck at max duration
simply make a text aura, use %p and apply a buff to yourself that has a duration

**Comment by Causese** [comment]
improt: https://wago.io/EJ-oJ_8tb
error: `Message: [string "--[[ Error in 'Combo Points - Feral' ]] ret..."]:3: attempt to perform arithmetic on field 'cloneId' (a string value)`
**line 3 in animation code probably causese the error**

**Comment by InfusOnWoW** [comment]
@Causese There's a change in that we now run the trigger functions with a special OPTIONS event from the options, so that TSU auras can adjust their states as they wish. That'll break some auras, and thus I've opted to hide errors from various places for that code path. I've added the same hiding of errors for animations now.

**Comment by InfusOnWoW** [comment]
I need to test the "in game update" problem. That is, if the user upgrades while wow is running SubText won't be loaded and I need to figure out what that means for the modernize code.

**Comment by InfusOnWoW** [comment]
And there was indeed a problem with not restarting where all texts would have been lost...

**Pull request merged**